index.bs

```diff
@@ -2725,6 +2769,284 @@ The [=remote end steps=] with <var ignore>session</var> and |command parameters|
 
 1. Return [=success=] with data null.
 
+</div>
+#### The browsingContext.locateNodes Command ####  {#command-browsingContext-locateNodes}
+
+The <dfn export for=commands>browsingContext.locateNodes</dfn> command returns a
+list of all nodes matching the specified locators.
+
+<dl>
+   <dt>Command Type</dt>
+   <dd>
+      <pre class="cddl remote-cddl">
+      browsingContext.LocateNodes = (
+        method: "browsingContext.locateNodes",
+        params: browsingContext.LocateNodesParameters
+      )
+
+      browsingContext.LocateNodesParameters = {
```

Comment:
do we need to add the ownership parameters to retain results?

Reply:
I don't know. Do we?

Reply:
I think we would need that if we want the user to make use of the returned handles. Currently, it looks like we just return a snapshot of the elements (I see resultOwnerhip set to none). We also probably want to think about realms and sandboxes (maybe not immediately in this proposal).

Reply:
@OrKoN I've added an optional `ownership` property to the command parameters. Does that alleviate this concern?

Reply:
Just a note that if this API needs to be compatible with Puppeteer (Playwright? and maybe others), we need a way to retain ownership because it is what Puppeteer element handling model assumes (and we use script.evaluate for that, that supports that). Also, if this API ends up being defined in terms of DOM APIs, having an option to specify realms seems useful since the results may differ across realms (e.g., the main realm might have DOM APIs overriden). It probably does not make sense to define this API in terms of DOM APIs though: I guess in such a  case one could use DOM APIs directly via scripts. Still we would likely need a way to control the result ownership.

Reply:
> if you wanted to find a node that was subsequently deleted

the node might get detached from DOM so having the protocol retain the object is useful, if, for example, you are testing for specific content of the element in subsequent calls and you don't care if it is connected to DOM at a later point. If the object is not retained, in highly dynamic apps it might require constant re-querying to do anything in multiple operations. Also, it might work or not depending on gc.

Reply:
@jgraham yeah, we have some examples in Chrome DevTools frontend e2e tests where UI updates (usually driven by the browser) are done by removing the content and inserting new DOM elements. References to the old elements are not kept. The test goes like this: 1) get an element 2) check its attributes or content (sometimes more complex stuff) via a second call. In this case, you either need to retry if the element the test managed to grab got GCed, or know really well that the element will not be removed by the next update (e.g., by preventing updates). Apart from that, our overall API in Puppeteer is that the element is retained unless you dispose it (alternatively, it is disposed by navigation).

We currently don't plan to use this API for Puppeteer anyway so the ownership capabilities can be also added later (if there is a need).

Reply:
@jgraham for `sharedId` we keep references to elements in weak sets but do not hold a strong reference to it. So if an element is removed from the DOM and got GC'ed the `sharedId` in comparison to the `handle` is no longer valid. That's reminds me that I should write a wdspec test for that!

Reply:
Folks, I'm going to resolve this conversation, as I've redefined the algorithm to only return a list of `sharedId`s. If anyone would rather see different behavior, I'm open to iterating on the proposal, but I warn I won't have a ton of time for revising this PR going forward. We're at the point where someone probably needs to take over authorship from me.

Reply:
Please note that my comment might not block getting the first iteration of this command landed. Having handling of ownership included could be a next step.